### PR TITLE
Enhance container builds and do vulnerability scans

### PR DESCRIPTION
**Pull request by JAORMX**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This uses docker instead of podman/buildah. With this, it adds extra
metadata to the container images. Finally, this introduces Trivy
to do vulnerability scans after building
images.

These scans can alert us to know when a tag bump is needed or when we're
misconfiguring an image.

This also resolves issues found by Trivy by updating golang dependencies.

#### Which issue(s) this PR fixes:

None

#### Does this PR have test?

The image build tests

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Review by pjbgf — APPROVED** [review]
/lgtm

**Comment by pjbgf** [comment]
Not sure whether the action supports a way in which it can access the image without using UNIX sockets:

```
FATAL	image scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:
	* unable to inspect the image (build): Error: No such image: build
	* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
	* containerd socket not found: /run/containerd/containerd.sock
	* GET https://index.docker.io/v2/library/build/manifests/latest: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:library/build Type:repository]]
```

**Comment by JAORMX** [comment]
@pjbgf if this doesn't work out, I can use Grype instead, it has an action that should work without sockets IIRC.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -62,29 +62,86 @@ jobs:
...
 
   ubi-image:
     runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
-      - uses: redhat-actions/buildah-build@d097e2e3d2a45b68a7c09040e3de33427bab66af
-        id: build-image
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
```
Comment (pjbgf) [inline .github/workflows/build.yml:111]: I think for most things we are using the hash commit as version:
```suggestion
        uses: docker/setup-buildx-action@312f67c6629946ba17269558b63d0bd84a63caab
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -62,29 +62,86 @@ jobs:
...
```
Comment (pjbgf) [inline .github/workflows/build.yml:67]:
```suggestion
        uses: docker/setup-buildx-action@312f67c6629946ba17269558b63d0bd84a63caab
```

**Review by pjbgf — CHANGES_REQUESTED** [review]
Let's align the action versions with the rest of the workflow which use hash version, then we can benefit from hermetic workflows.

Dependabot will keep this updated for us.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: |
+            quay.io/security-profiles-operator/security-profiles-operator
+          tags: |
+            type=ref,event=branch
+            type=sha,format=long
+      - name: Build (and push if needed)
+        uses: docker/build-push-action@v3
```
Comment (pjbgf) [inline .github/workflows/build.yml:85]:
```suggestion
        uses: docker/build-push-action@965c6a410d446a30e95d35052c67d6eded60dad6
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+        with:
+          context: .
+          file: Dockerfile.build-image
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          push: ${{ github.ref == 'refs/heads/main' }}
+          load: true
+      - name: Run container image vulnerability scanner
+        uses: aquasecurity/trivy-action@master
```
Comment (pjbgf) [inline .github/workflows/build.yml:94]:
```suggestion
        uses: aquasecurity/trivy-action@d63413b0a4a4482237085319f7f4a1ce99a8f2ac
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -62,29 +62,86 @@ jobs:
...
 
   ubi-image:
     runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
-      - uses: redhat-actions/buildah-build@d097e2e3d2a45b68a7c09040e3de33427bab66af
-        id: build-image
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
```
Comment (pjbgf) [inline .github/workflows/build.yml:111]:
```suggestion
        uses: docker/setup-buildx-action@dc7b9719a96d48369863986a06765841d7ea23f6
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+      - name: Login to Quay
+        uses: docker/login-action@v2
```
Comment (pjbgf) [inline .github/workflows/build.yml:113]:
```suggestion
        uses: docker/login-action@49ed152c8eca782a232dede0303416e8f356c37b
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+        with:
+          registry: quay.io/security-profiles-operator
+          username: security-profiles-operator+github
+          password: ${{ secrets.QUAY_TOKEN }}
+        if: github.ref == 'refs/heads/main'
+      - name: Image metadata
+        id: meta
+        uses: docker/metadata-action@v4
```
Comment (pjbgf) [inline .github/workflows/build.yml:121]:
```suggestion
        uses: docker/metadata-action@69f6fc9d46f2f8bf0d5491e4aabe0bb8c6a4678a
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+          images: |
+            quay.io/security-profiles-operator/security-profiles-operator
+          flavor: |
+            suffix=-ubi
+          tags: |
+            type=ref,event=branch
+            type=sha,format=long
+      # TODO(jaosorior): Push UBI image too
+      - name: Build
+        uses: docker/build-push-action@v3
```
Comment (pjbgf) [inline .github/workflows/build.yml:132]:
```suggestion
        uses: docker/build-push-action@c84f38281176d4c9cdb1626ffafcd6b3911b5d94
```

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -62,29 +62,86 @@ jobs:
...
+        with:
+          context: .
+          file: Dockerfile.ubi
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+          load: true
+      - name: Run container image vulnerability scanner
+        uses: aquasecurity/trivy-action@master
```
Comment (pjbgf) [inline .github/workflows/build.yml:140]:
```suggestion
        uses: aquasecurity/trivy-action@d63413b0a4a4482237085319f7f4a1ce99a8f2ac
```

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -62,29 +62,86 @@ jobs:
     runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
-      - uses: redhat-actions/buildah-build@d097e2e3d2a45b68a7c09040e3de33427bab66af
-        id: build-image
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@312f67c6629946ba17269558b63d0bd84a63caab
+      - name: Login to Quay
+        uses: docker/login-action@v2
```
Comment (pjbgf) [inline .github/workflows/build.yml:69]:
```suggestion
        uses: docker/login-action@d398f07826957cd0a18ea1b059cf1207835e60bc
```

**Review by pjbgf — COMMENTED** [review]
There are a few more actions that we can set their hash versions.

**Review by pjbgf — APPROVED** [review]
/lgtm

great job @JAORMX, thanks for working on this. :bow:

**Pull request merged**